### PR TITLE
wallet: Add wallet method to detect if a key is "active"

### PR DESCRIPTION
**Pull request by pinheadmz**
This is a first step towards closing https://github.com/bitcoin/bitcoin/issues/3314 by implementing functions in Wallet and ScriptPubKeyMan that determine if a key is "active" meaning "derived from a currently active scriptpubkey manager." 

A key is NOT "active" if:
- It was imported using an `import{privkey, address, pubkey}` rpc
- It was derived before the wallet was encrypted (which replaces the active SPKM(s))

Users should be informed that these keys may be compromised:
- Some other wallet generated the private key and may still have a copy
- The private key was stored on disk unencrypted

This PR adds a new field `isactive` to rpc `getaddressinfo`, follow-up work will be adding this to the GUI address book as well, [something like this](https://github.com/bitcoin/bitcoin/issues/3314#issuecomment-81567665). The original issue also recommends sweeping all value from keys in this category, I think that feature can be discussed in the future as a follow-up.

This PR also patches `PKDescriptor` from https://github.com/bitcoin/bitcoin/pull/22051 where matching public keys were not returned.

Follow-up PR for the GUI: https://github.com/bitcoin-core/gui/pull/723

### CODE REVIEW

File: src/wallet/test/scriptpubkeyman_tests.cpp [hunk 1]
```diff
@@ -39,5 +40,146 @@ BOOST_AUTO_TEST_CASE(CanProvide)
...
+    BOOST_CHECK(scripts2.size() == 1000);
+
+    // Use key that is not the next key
+    // (i.e. address gap in wallet recovery)
+    {
+        LOCK(spkm->cs_desc_man);
+        WalletDescriptor descriptor = spkm->GetWalletDescriptor();
+        FlatSigningProvider provider;
+        std::vector<CScript> scripts3;
+        descriptor.descriptor->ExpandFromCache(/*index=*/10, descriptor.cache, scripts3, provider);
```
Comment (fanquake) [inline src/wallet/test/scriptpubkeyman_tests.cpp:142]: https://cirrus-ci.com/task/6502838134636544:
```bash
wallet/test/scriptpubkeyman_tests.cpp:142:48: error: argument name 'index' in comment does not match parameter name 'pos' [bugprone-argument-comment,-warnings-as-errors]
        descriptor.descriptor->ExpandFromCache(/*index=*/10, descriptor.cache, scripts3, provider);
                                               ^
./script/descriptor.h:137:38: note: 'pos' declared here
    virtual bool ExpandFromCache(int pos, const DescriptorCache& read_cache, std::vector<CScript>& output_scripts, FlatSigningProvider& out) const = 0;
```

**Comment by fanquake** [comment]
https://github.com/bitcoin/bitcoin/pull/27216/checks?check_run_id=11804958852
```bash
Assertion failed: lock cs_wallet not held in ./wallet/wallet.h:444; locks held:
'cs_KeyStore' in wallet/scriptpubkeyman.cpp:1185 (in thread 'test')
```

File: src/wallet/test/scriptpubkeyman_tests.cpp (hunk 1 above)
Comment (pinheadmz) [reply src/wallet/test/scriptpubkeyman_tests.cpp:142]: oooh thanks, didn't realize those comments were parsed by anything!

**Comment by pinheadmz** [comment]
rebased on master and addressed some nits from follow-up PR https://github.com/bitcoin-core/gui/pull/723

**Review by jonatack — COMMENTED** [review]
Concept ACK, tested and reviewed each commit. Some thoughts and questions follow.

> This PR also patches PKDescriptor from https://github.com/bitcoin/bitcoin/pull/22051 where matching public keys were not returned.

I didn't see where/how; can you help or explain?

File: src/script/descriptor.cpp [hunk 2]
```diff
@@ -730,8 +730,11 @@ class PKDescriptor final : public DescriptorImpl
 private:
     const bool m_xonly;
 protected:
-    std::vector<CScript> MakeScripts(const std::vector<CPubKey>& keys, Span<const CScript>, FlatSigningProvider&) const override
+    std::vector<CScript> MakeScripts(const std::vector<CPubKey>& keys, Span<const CScript>, FlatSigningProvider& out) const override
     {
+        CKeyID id = keys[0].GetID();
+        out.pubkeys.emplace(id, keys[0]);
```
[the remark this answers is not in the document]
Comment (pinheadmz) [reply src/script/descriptor.cpp:736]: The style here matches the neighboring descriptors but I see your point about using `.at(0)` -- do you think all the `descriptor::MakeScripts()` should be cleaned up like this?

Also this chunk is the patch alluded to in the comment
> This PR also patches PKDescriptor from https://github.com/bitcoin/bitcoin/pull/22051 where matching public keys were not returned.

When we expand a bare pubkey descriptor we weren't populating `out.pubkeys` and so `GetAffectedKeys()` wouldn't return anything

File: src/wallet/scriptpubkeyman.h [hunk 3]
```diff
@@ -198,6 +198,9 @@ class ScriptPubKeyMan
      */
     virtual std::vector<WalletDestination> MarkUnusedAddresses(const CScript& script) { return {}; }
 
+    /* Determines if address is derived from active key manager */
+    virtual bool IsKeyActive(const CScript& script) { return false; };
```
[the remark this answers is not in the document]
Comment (pinheadmz) [reply src/wallet/scriptpubkeyman.h:202]: thanks will add this -- also I think these could be `[[nodiscard]]` as well? But I don't see that for any of the other `bool Is...()` members in SPKM...

File: src/wallet/scriptpubkeyman.cpp [hunk 4]
```diff
@@ -392,6 +392,28 @@ std::vector<WalletDestination> LegacyScriptPubKeyMan::MarkUnusedAddresses(const
...
+    for (const auto& key_id : GetAffectedKeys(script, *this)) {
+        auto it = mapKeyMetadata.find(key_id);
+        if (it == mapKeyMetadata.end()) {
+            // This key must be really old
+            return false;
+        }
+        const CKeyMetadata& meta = it->second;
+        if (m_hd_chain.seed_id == meta.hd_seed_id)
+            return true;
+    }
```
[the remark this answers is not in the document]
Comment (pinheadmz) [reply src/wallet/scriptpubkeyman.cpp:411]: Thanks, updating.

File: src/wallet/rpc/addresses.cpp [hunk 5]
```diff
@@ -602,6 +603,7 @@ RPCHelpMan getaddressinfo()
     ret.pushKVs(detail);
 
     ret.pushKV("ischange", ScriptIsChange(*pwallet, scriptPubKey));
+    ret.pushKV("isactive", pwallet->IsDestinationActive(dest));
```
[the remark this answers is not in the document]
Comment (pinheadmz) [reply src/wallet/rpc/addresses.cpp:606]: ok, moving `isactive` next to `ismine`. You have a good point about legacy wallet deprecation. I suppose when that happens most of this new code will be deprecated as well since its mainly implemented only in `LegacyScriptPubKeyMan`. Would it make sense to only add the field to RPC response in legacy wallets?

File: src/wallet/test/scriptpubkeyman_tests.cpp [hunk 6]
```diff
@@ -39,5 +40,159 @@ BOOST_AUTO_TEST_CASE(CanProvide)
...
+    // Generate 20 keypool keys (10 internal, 10 external)
+    {
+        LOCK(wallet.cs_wallet);
+        spkm.SetupGeneration();
+    }
+
+    // 4 scripts per keypool key (P2PK, P2PKH, P2WPKH, P2SH-P2WPKH)
+    // Plus 4 scripts for the seed key
+    auto scripts1 = spkm.GetScriptPubKeys();
+    BOOST_CHECK(scripts1.size() == 84);
```
[the remark this answers is not in the document]
Comment (pinheadmz) [reply src/wallet/test/scriptpubkeyman_tests.cpp:65]: thanks, I do prefer!

File: src/wallet/test/scriptpubkeyman_tests.cpp [hunk 7]
```diff
@@ -39,5 +40,159 @@ BOOST_AUTO_TEST_CASE(CanProvide)
...
+        BOOST_CHECK(spkm->IsKeyActive(script));
+    }
+
+    // ...but at the wallet level all the keys from that SPKM are deactivated
+    int num_script_keys_not_found = 0;
+    for (const CScript& script : scripts4) {
+        WitnessV0ScriptHash scripthash(script);
+        if (!wallet.IsDestinationActive(scripthash))
+            num_script_keys_not_found++;
+    }
```
[the remark this answers is not in the document]
Comment (pinheadmz) [reply src/wallet/test/scriptpubkeyman_tests.cpp:193]: thanks got it

File: src/wallet/wallet.cpp [hunk 8]
```diff
@@ -2557,6 +2557,16 @@ void CWallet::ForEachAddrBookEntry(const ListAddrBookFunc& func) const
     }
 }
 
+bool CWallet::IsDestinationActive(const CTxDestination& dest) const
+{
+    CScript script = GetScriptForDestination(dest);
```
[the remark this answers is not in the document]
Comment (pinheadmz) [reply src/wallet/wallet.cpp:2562]: ah really cool with `any_of` thanks

File: test/functional/wallet_keypool.py [hunk 9]
```diff
@@ -4,32 +4,107 @@
...
+        import_priv_data = nodes[0].getaddressinfo("bcrt1qa985v5d53qqtrfujmzq2zrw3r40j6zz4ns02kj")
+        assert import_priv_data["ismine"]
+        assert not import_priv_data["iswatchonly"]
+        assert not import_priv_data["isactive"]
+
         # Encrypt wallet and wait to terminate
         nodes[0].encryptwallet('test')
+        addr9_after_encrypting_data = nodes[0].getaddressinfo(addr9)
+        # Key is from unencrypted seed, no longer considered active
+        assert addr9_after_encrypting_data['isactive'] == False
```
[the remark this answers is not in the document]
Comment (pinheadmz) [reply test/functional/wallet_keypool.py:106]: okay yes

File: src/wallet/rpc/addresses.cpp [hunk 10]
```diff
@@ -506,7 +506,8 @@ RPCHelpMan getaddressinfo()
                         {RPCResult::Type::STR, "desc", /*optional=*/true, "A descriptor for spending coins sent to this address (only when solvable)."},
                         {RPCResult::Type::STR, "parent_desc", /*optional=*/true, "The descriptor used to derive this address if this is a descriptor wallet"},
                         {RPCResult::Type::BOOL, "isscript", "If the key is a script."},
-                        {RPCResult::Type::BOOL, "ischange", "If the address was used for change output."},
+                        {RPCResult::Type::BOOL, "ischange", "If the address was or will be used for change output."},
```
[the remark this answers is not in the document]
Comment (pinheadmz) [reply src/wallet/rpc/addresses.cpp:509]: ok?

> "If the address is reserved for use as a change output. It may or may not have already been used as change."

**Comment by pinheadmz** [comment]
@jonatack thanks so much for the great review, addressed comments (1 or 2 questions along the way) rebase on master and also rebased the GUI follow up

**Review by jonatack — COMMENTED** [review]
ACK with some suggestions below.

The release note and added test coverage could optionally all be in separate commits, provided each commit is hygienic, i.e. builds and all tests pass. It's a bit easier for reviewers if updates needed for tests to pass with a change are in the same commit, while additional coverage is in separate commits.

File: src/wallet/scriptpubkeyman.cpp [hunk 11]
```diff
@@ -392,6 +392,23 @@ std::vector<WalletDestination> LegacyScriptPubKeyMan::MarkUnusedAddresses(const
     return result;
 }
 
+[[nodiscard]] bool LegacyScriptPubKeyMan::IsKeyActive(const CScript& script) const
```
[the remark this answers is not in the document]
Comment (pinheadmz) [reply src/wallet/scriptpubkeyman.cpp:395]: ah yeah "...Appears in a function declaration, enumeration declaration, or class declaration. "

but theres a few leaks in bitcoin...
```
--> git grep nodiscard *.cpp | wc -l
       8
```

**Comment by pinheadmz** [comment]
thanks again @jonatack 
rebase to 8139075244f1f361a14f80920c45d8c4f41d553e
- refactor commits to separate tests in hygenic order!
- address review nits

**Review by jonatack — COMMENTED** [review]
ACK 8139075244f1f361a14f80920c45d8c4f41d553e modulo eyes from someone more familiar with the wallet to validate the approach.

Commit 2a5774f45c07f0d000f0cbdf0 should probably be renamed to something like `test: cover ScriptPubKeyMan::IsKeyActive and IsDestinationActive`

Re-reviewing this shuffled version was a little trickier due to the unneeded rebase. Please rebase only when you have to. Aside from that, the reorganized commits seem easier to review now.

File: src/wallet/scriptpubkeyman.h [hunk 12]
```diff
@@ -198,6 +198,9 @@ class ScriptPubKeyMan
      */
     virtual std::vector<WalletDestination> MarkUnusedAddresses(const CScript& script) { return {}; }
 
+    /* Determines if address is derived from active key manager */
+    [[nodiscard]] virtual bool IsKeyActive(const CScript& script) const { return false; };
```
[the remark this answers is not in the document]
Comment (pinheadmz) [reply src/wallet/scriptpubkeyman.h:202]: Ah thanks for the ack, I'll update this and the commit message you mentioned if I need to revise again.

**Comment by pinheadmz** [comment]
> Needs rebase

done thanks

File: src/wallet/scriptpubkeyman.h (hunk 12 above)
[the remark this answers is not in the document]
Comment (pinheadmz) [reply src/wallet/scriptpubkeyman.h]: Thanks, both comments addressed with rebase

**Comment by luke-jr** [comment]
I don't mind the new field, but it's not correct to say users should be discouraged from using any key they import...

**Review by jonatack — COMMENTED** [review]
re-ACK f7cd46280fefa06f9487a8a2927de093af1d421d only changes since my last review are rebase, making the `IsKeyActive` virtual function declaration pure, and improving a commit message

- modulo https://github.com/bitcoin/bitcoin/pull/27216#issuecomment-1527773928
- happy to re-review if any changes are needed for https://github.com/bitcoin/bitcoin/pull/27216#issuecomment-1603302152

**Comment by jarolrod** [comment]
Concept ACK

**Comment by pinheadmz** [comment]
> I don't mind the new field, but it's not correct to say users should be discouraged from using any key they import...

@luke-jr  Fair point, I updated the OP (no code changes needed)

>  legacy wallet support may be dropped in the next couple releases.

@jonatack It may take about that long for this PR to get merged anyway 🤣. So, I can remove the legacy stuff from this PR right now and only descriptor wallet users can enjoy the feature, OR, we can keep `isactive` in there for now and then remove it along with the rest of the legacy wallet stuff when that is appropriate. Thoughts? Preference?

File: src/wallet/rpc/addresses.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (pinheadmz) [reply src/wallet/rpc/addresses.cpp]: > I'm not sure if or when legacy wallet support is expected to be phased out, but if isactive is only different from ismine in legacy wallets, is it worth adding a field (and all the new methods), versus just updating the ismine field documentation in the help for descriptor wallets? (When I'm double-checking an address with getaddressinfo before using it to receive, ismine is a field I'm already checking.)

@jonatack after diving back in to this issue I realized that `ismine != isactive` for descriptor wallets all the time. I added tests in `wallet_keypool` to confirm this. The difference is whether or not the SPKM is ACTIVE (A descriptor wallet can have both active and inactive SPKMs and `ismine` at the RPC level merely indicates if the wallet can spend from the address).

So to summarize the test I just added:
- create descriptor wallet
- getnewaddress, getaddressinfo: ismine and isactive are both true
- encrypt the wallet
- getaddressinfo again: ismine is true but isactive is false

**Comment by S3RK** [comment]
Unfortunately, we can't use "activeness" of a descriptor to determine if it was generated before encryption. 

Descriptor wallets allow as many descriptors as you want, but only one descriptor per OutputType could be active at a time. The activeness only specifies which descriptor to use when generating new addresses for that type. Therefore, there are more scenarios, besides the two you've listed, in which a key will be NOT "active". The simplest one would be — I just imported two p2wpkh descriptors for two BIP-44 accounts that I have. And only one of these descriptors could be active at a time.

In general, I believe that the proper solution to "exposed" keys is to get a new wallet with new set of keys, rather than rotating keys in existing wallet. We don't have a good separation between UTXOs/addresses within a wallet. On the other hand we already support multiple wallets, which are completely isolated so there are no chances of leaking/mixing UTXOs and addresses.

**Review by S3RK — COMMENTED** [review]
Unfortunately, we can't use "activeness" of a descriptor to determine if it was generated before encryption.

Descriptor wallets allow as many descriptors as you want, but only one descriptor per OutputType could be active at a time. The activeness only specifies which descriptor to use when generating new addresses for that type. Therefore, there are more scenarios, besides the two you've listed, in which a key will be NOT "active". The simplest one would be — I just imported two p2wpkh descriptors for two BIP-44 accounts that I have. And only one of these descriptors could be active at a time.

In general, I believe that the proper solution to "exposed" keys is to get a new wallet with new set of keys, rather than rotating keys in existing wallet. We don't have a good separation between UTXOs/addresses within a wallet. On the other hand we already support multiple wallets, which are completely isolated so there are no chances of leaking/mixing UTXOs and addresses.

**Comment by pinheadmz** [comment]
> I just imported two p2wpkh descriptors for two BIP-44 accounts that I have. And only one of these descriptors could be active at a time.

You would have to set `"active":"true"` when executing RPC `importdescriptors` and the default is `false`. But I see your point, considering this feature is for safety its not all that difficult for an advanced user to bypass that check.

Even if you did do this (import with active:true), I still think it is useful for getaddrinfo to notify you if an address is derived from an inactive descriptor. The issue I am targeting with this PR is more about protecting less advanced users (particularly with the GUI) from giving out "risky" receive addresses.

**Comment by S3RK** [comment]
I have two concerns with the approach:
* I don't like changing semantic of "active" field  to mean it's somehow unsafe.
* Even for less advances users, I don't think it fully addresses the describe use-case, because it doesn't allow easy sweeping of compromised keys.

1/ The activeness indicates from which descriptor to derive new addresses. Based on the semantics of that field, It is incorrect to imply that addresses from inactive descriptors are less safe that addresses from active descriptors. Moreover, activeness of the descriptor is a user configurable parameter and we can't just reuse it to store information about "safety" of the descriptor. 

2/ IIUC here's the use cases this PR tries to address: a user believes existing private keys are compromised and rotates the keys in the wallet. Since the keys are compromised it's unsafe to receive new and keep old funds. So after rotating the keys the user would have to sweep all the compromised keys. If I'm not mistaken, we currently don't support sweeping coins from a descriptor. However, we support the whole wallet sweeping with `sendall` command.

Separate wallet solution is also less error prone because it doesn't rely on a user to pay attention to the addresses to maintain safety. This approach (with optional sweeping) also works for privacy leaks, because it also prevents mixing UTXO histories.

**Review by pablomartin4btc — COMMENTED** [review]
Concept ACK
Interesting points of view from both @pinheadmz and @S3RK, regarding "activeness" concept for descriptor wallets, the use case scenario this PR tries to solve and safety related with rotating keys vs wallet sweeping.
I'll review this PR further.

**Comment by pinheadmz** [comment]
I'm closing this to rethink the approach. Since 'active' descriptors can be imported it defeats the value of that field for this purpose. Really it seems the only property the wallet can really know for sure is if it generated keys into an encrypted wallet. I'll be back with a UI indicator for keys that we know have always been "safe".
